platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy

```diff
@@ -263,7 +239,33 @@ class SoftwareTypeDeclarationIntegrationTest extends AbstractIntegrationSpec imp
 
         then:
         failure.assertHasCause("Failed to apply plugin 'com.example.test-software-type'.")
-        failure.assertHasCause("A plugin with type 'org.gradle.test.SoftwareTypeImplPlugin' was registered as a software type plugin, but it does not expose any software types. Software type plugins must expose software types via properties with the @SoftwareType annotation.")
+        failure.assertHasCause("A problem was found with the SoftwareTypeImplPlugin plugin.")
+        failure.assertHasCause("Type 'org.gradle.test.SoftwareTypeImplPlugin' is registered as a software type plugin but does not expose a software type.")
+    }
+
+    def 'sensible error when a software type plugin is registered that exposes multiple software types'() {
+        given:
+        withSoftwareTypePluginThatExposesMultipleSoftwareTypes().prepareToExecute()
+
+        file("settings.gradle.dcl") << pluginsFromIncludedBuild
+
+        file("build.gradle.dcl") << declarativeScriptThatConfiguresOnlyTestSoftwareType + """
+            anotherSoftwareType {
+                id = "test2"
+
+                foo {
+                    bar = "fizz"
+                }
+            }
+        """
+
+        when:
+        fails(":printTestSoftwareTypeExtensionConfiguration")
+
+        then:
+        failure.assertHasCause("Failed to apply plugin 'com.example.test-software-type'.")
+        failure.assertHasCause("A problem was found with the SoftwareTypeImplPlugin plugin.")
+        failure.assertHasCause("Type 'org.gradle.test.SoftwareTypeImplPlugin' is registered as a software type plugin, but it exposes multiple software types.")
```

Comment:
🤔 Can we expand this message to note the names of software types declared?  Something like:

```
Plugin with type 'com.example.restricted.RestrictedPlugin' is registered as a software type plugin, but it declares more than one software type: application, library, utility. Software type plugins must expose only a single software type.
```

Alternately, as the exception already implements `ResolutionProvider`, we could add a suggestion to expose only a single software type or register multiple plugins.

Reply:
Yep.  I think that's a little better.